cdap-data-fabric/src/main/java/co/cask/cdap/spi/metadata/dataset/DatasetMetadataStorage.java

```diff
@@ -349,62 +350,104 @@ private Metadata read(MetadataDatasetContext context, Read read) {
 
   @Override
   public SearchResponse search(SearchRequest request) {
-    NamespaceId namespace = null;
-    Set<EntityScope> entityScopes;
-    if (request.getNamespaces() == null || request.getNamespaces().isEmpty()) {
-      entityScopes = EnumSet.allOf(EntityScope.class);
-    } else {
-      boolean hasSystem = false;
-      for (String ns : request.getNamespaces()) {
-        if (ns.equals(NamespaceId.SYSTEM.getNamespace())) {
-          hasSystem = true;
-        } else {
-          if (namespace != null) {
-            throw new UnsupportedOperationException(String.format(
-              "This implementation supports at most one non-system namespace, but %s as well as %s were given",
-              namespace.getNamespace(), ns));
-          }
-          namespace = new NamespaceId(ns);
-        }
-      }
-      if (hasSystem) {
-        if (namespace == null) {
-          namespace = NamespaceId.SYSTEM;
-          entityScopes = Collections.singleton(EntityScope.SYSTEM);
-        } else {
-          entityScopes = EnumSet.allOf(EntityScope.class);
-        }
-      } else {
-        entityScopes = Collections.singleton(EntityScope.USER);
-      }
-    }
+    ImmutablePair<NamespaceId, Set<EntityScope>> namespaceAndScopes =
+      determineNamespaceAndScopes(request.getNamespaces());
+    ImmutablePair<String, int[]> cursorOffsetAndLimits = determineCursorOffsetAndLimits(request);
```

Comment:
A class with 5 fields would make this easier to understand than a string and an int array.

Reply:
yeah you're right. I was trying to avoid creating another class just for this. I really wish Java had a generic Tuple class. Will change it